crates/ide-diagnostics/src/handlers/missing_unsafe.rs

```diff
@@ -9,11 +16,60 @@ pub(crate) fn missing_unsafe(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsaf
         "this operation is unsafe and requires an unsafe function or block",
         ctx.sema.diagnostics_display_range(d.expr.clone().map(|it| it.into())).range,
     )
+    .with_fixes(fixes(ctx, d))
+}
+
+fn fixes(ctx: &DiagnosticsContext<'_>, d: &hir::MissingUnsafe) -> Option<Vec<Assist>> {
+    let root = ctx.sema.db.parse_or_expand(d.expr.file_id)?;
+    let expr = d.expr.value.to_node(&root);
+
+    let node_to_add_unsafe_block = pick_best_node_to_add_unsafe_block(ctx, &expr);
+
+    let replacement = format!("unsafe {{ {} }}", node_to_add_unsafe_block.text());
+    let edit = TextEdit::replace(node_to_add_unsafe_block.text_range(), replacement);
+    let source_change =
+        SourceChange::from_text_edit(d.expr.file_id.original_file(ctx.sema.db), edit);
+    Some(vec![fix("add_unsafe", "Add unsafe block", source_change, expr.syntax().text_range())])
+}
+
+// Find the let statement or expression statement closest to the `expr` in the
+// ancestor chain.
+//
+// Why don't we just add an unsafe block around the `expr`?
+//
+// Consider this example:
+// ```
+// STATIC_MUT += 1;
+// ```
+// We can't add an unsafe block to the left-hand side of an assignment.
+// ```
+// unsafe { STATIC_MUT } += 1;
+// ```
+//
+// Or this example:
+// ```
+// let z = STATIC_MUT.a;
+// ```
+// We can't add an unsafe block like this:
+// ```
+// let z = unsafe { STATIC_MUT } .a;
+// ```
+fn pick_best_node_to_add_unsafe_block(
+    ctx: &DiagnosticsContext<'_>,
+    expr: &ast::Expr,
+) -> SyntaxNode {
+    let Some(let_or_expr_stmt) = ctx.sema.ancestors_with_macros(expr.syntax().clone()).find(|node| {
```

Comment:
We need to be sure not to suggest fixes for when this diagnostic comes from a macro, so we should instead do something like
```rs
ctx.sema.original_ast_node(expr).syntax().ancestors()...
```